pyqtorch/embed.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from logging import getLogger
+from typing import Any, Callable
+
+import torch
+
+logger = getLogger(__name__)
+
+
+def torch_call(
+    abstract_fn: str, args: list[str | float]
+) -> Callable[[dict], torch.Tensor]:
+    """Convert a `Call` object into a torchified function which can be evaluated using
+    a vparams and inputs dict.
+    """
+    fn = getattr(torch, abstract_fn)
+
+    def evaluate(inputs: dict[str, torch.Tensor]) -> torch.Tensor:
+        tensor_args = []
+        for symbol in args:
+            if isinstance(symbol, float):
+                tensor_args.append(torch.tensor(symbol))
+            elif isinstance(symbol, str):
+                tensor_args.append(inputs[symbol])
+        return fn(*tensor_args)
+
+    return evaluate
+
+
+class Embedding(torch.nn.Module):
+    """
+    vparam_names: A list of abstract variational parameter names.
+    fparam_names: A list of abstract feature parameter names.
+    leaf_to_call: Map from intermediate and leaf variables (which will be used as angles in gates)
+                to torch callables which expect a dictionary of root and intermediate variables.
+    dtype: The precision of the parameters.
+    device: The device on which the parameters are stored.
+    """
+
+    def __init__(
+        self,
+        vparam_names: list[str],
+        fparam_names: list[str],
+        leaf_to_call: dict[str, Callable],
+        dtype: torch.dtype = torch.float64,
+        device: torch.device = torch.device("cpu"),
+    ) -> None:
+        super().__init__()
+        self.vparams = torch.nn.ParameterDict(
+            {p: torch.rand(1, requires_grad=True) for p in vparam_names}
+        )
+        self.fparams: dict[str, torch.Tensor | None] = {p: None for p in fparam_names}
+        self.leaf_to_call: dict[str, Callable] = leaf_to_call
```

Comment:
Naming suggestion: `torchify_map` ?

Reply:
outdated since it also allows other engines than torch